optashift-employee-rostering-gwtui/src/main/java/org/optaplanner/openshift/employeerostering/gwtui/client/employee/EmployeeSubform.java

```diff
@@ -0,0 +1,176 @@
+        return new Promise<Employee>((res, rej) -> {
+            Set<ConstraintViolation<Employee>> validationErrorSet = validate();
+            if (validationErrorSet.isEmpty()) {
+                res.onInvoke(getNewValue());
+            } else {
+                rej.onInvoke(validationErrorSet);
+            }
+        });
+
+    }
+
+    @Override
+    public void onUpdate(Map<String, Skill> data) {
+        skillProficiencySet.clear();
+        data.forEach((name, skill) -> {
+            Option option = new Option();
+            option.setName(name);
+            option.setValue(name);
+            option.setText(name);
+            skillProficiencySet.add(option);
+        });
+        skillProficiencySet.refresh();
+    }
+
+    @Override
+    protected void deleteRow() {
+        EmployeeRestServiceBuilder.removeEmployee(tenantStore.getCurrentTenantId(), getValue().getId(),
+                FailureShownRestCallback.onSuccess(success -> {
+                    dataInvalidationEvent.fire(new DataInvalidation<Employee>());
+                }));
+    }
+
+    @Override
+    protected void updateRow() {
+        getIfValid()
+                .then((e) -> {
+                    commitChanges();
+                    setEditing(false);
+                    EmployeeRestServiceBuilder.updateEmployee(tenantStore.getCurrentTenantId(), getNewValue(),
+                            FailureShownRestCallback.onSuccess(employee -> {
+                                dataInvalidationEvent.fire(new DataInvalidation<Employee>());
+                            }));
+                    return PromiseUtils.resolve();
+                })
+                .catch_((s) -> {
+                    @SuppressWarnings("unchecked")
+                    Set<ConstraintViolation<Employee>> errors = (Set<ConstraintViolation<Employee>>) s;
+                    ErrorPopup.show(CommonUtils.delimitCollection(errors, (e) -> e.getMessage(), "\n"));
+                    return PromiseUtils.resolve();
+                });
+    }
+
+    @Override
+    protected void createRow() {
+        getIfValid()
+                .then((e) -> {
+                    commitChanges();
+                    setEditing(false);
+                    EmployeeRestServiceBuilder.addEmployee(tenantStore.getCurrentTenantId(), getNewValue(),
+                            FailureShownRestCallback.onSuccess(employee -> {
+                                dataInvalidationEvent.fire(new DataInvalidation<Employee>());
+                            }));
+                    return PromiseUtils.resolve();
+                })
+                .catch_((s) -> {
+                    @SuppressWarnings("unchecked")
+                    Set<ConstraintViolation<Employee>> errors = (Set<ConstraintViolation<Employee>>) s;
+                    ErrorPopup.show(CommonUtils.delimitCollection(errors, (e) -> e.getMessage(), "\n"));
+                    return PromiseUtils.resolve();
+                });
+    }
```

Comment:
Duplicated from `updateRow()`. Please remove the `getIfValid()` method and make the validation check directly. Promises are not intended to be used as fancy try-catch blocks, although its API gives that idea :P